File: pkg/scanner/power_upgrades.go
package scanner

import (
	"bufio"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/zanicool/apex-cli/pkg/crawler"
	"github.com/zanicool/apex-cli/pkg/engine"
	"github.com/zanicool/apex-cli/pkg/oob"
)

// --- Payload Loader ---

var (
	sqliPayloadsFile []string
	xssPayloadsFile  []string
	paramWordlist    []string
	payloadsLoaded   bool
)

func loadPayloads() {
	if payloadsLoaded {
		return
	}
	payloadsLoaded = true
	// Find payloads directory relative to binary
	dirs := []string{"./payloads", "../payloads", "/usr/share/apex-cli/payloads"}
	for _, dir := range dirs {
		if _, err := os.Stat(dir); err == nil {
			sqliPayloadsFile = loadFile(filepath.Join(dir, "sqli.txt"))
			xssPayloadsFile = loadFile(filepath.Join(dir, "xss.txt"))
			paramWordlist = loadFile(filepath.Join(dir, "params.txt"))
			break
		}
	}
}

func loadFile(path string) []string {
	f, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer f.Close()
	var lines []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" && !strings.HasPrefix(line, "#") {
			lines = append(lines, line)
		}
	}
	return lines
}

// --- Deep JS Endpoint Extractor ---

func scanJSEndpoints(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	var mu sync.Mutex
	sem := make(chan struct{}, cfg.Threads)
	var wg sync.WaitGroup

	// Patterns that extract API routes from JS
	patterns := []*regexp.Regexp{
		regexp.MustCompile(`(?:"|')(/api/[a-zA-Z0-9_/\-\.]+)(?:"|')`),
		regexp.MustCompile(`(?:"|')(/v[0-9]+/[a-zA-Z0-9_/\-\.]+)(?:"|')`),
		regexp.MustCompile(`(?:fetch|axios\.\w+|\.get|\.post|\.put|\.delete|\.patch)\s*\(\s*(?:"|'|` + "`" + `)([^"'` + "`" + `\s]+)`),
		regexp.MustCompile(`(?:"|')(https?://[^"'\s]{10,})(?:"|')`),
		regexp.MustCompile(`(?:path|route|endpoint|url)\s*[:=]\s*(?:"|')([^"']+)(?:"|')`),
		regexp.MustCompile(`(?:"|')(/[a-zA-Z0-9_\-]+/[a-zA-Z0-9_\-]+(?:/[a-zA-Z0-9_\-]+)*)(?:"|')`),
	}

	// Collect all JS URLs
	jsURLs := make(map[string]bool)
	for _, page := range crawl.Pages {
		if strings.HasSuffix(page.URL, ".js") || strings.Contains(page.URL, "/static/js/") || strings.Contains(page.URL, "/_next/") || strings.Contains(page.URL, "/bundle") || strings.Contains(page.URL, "/chunk") {
			jsURLs[page.URL] = true
		}
	}
	// Add common JS paths
	if len(crawl.Pages) > 0 {
		base := strings.Split(crawl.Pages[0].URL, "/")[0] + "//" + strings.Split(crawl.Pages[0].URL, "/")[2]
		for _, p := range []string{"/main.js", "/app.js", "/bundle.js", "/vendor.js", "/chunk.js", "/_next/static/chunks/main.js", "/static/js/main.js", "/assets/index.js", "/dist/app.js"} {
			jsURLs[base+p] = true
		}
	}

	discoveredEndpoints := make(map[string]bool)
	for jsURL := range jsURLs {
		wg.Add(1)
		go func(u string) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()

			resp := http.Get(u)
			if resp.Err != nil || resp.StatusCode != 200 || len(resp.Body) < 100 {
				return
			}

			for _, pattern := range patterns {
				matches := pattern.FindAllStringSubmatch(resp.Body, -1)
				for _, m := range matches {
					if len(m) < 2 {
						continue
					}
					endpoint := m[1]
					if len(endpoint) < 3 || len(endpoint) > 200 {
						continue
					}
					// Skip static assets
					if strings.HasSuffix(endpoint, ".css") || strings.HasSuffix(endpoint, ".png") || strings.HasSuffix(endpoint, ".jpg") || strings.HasSuffix(endpoint, ".svg") || strings.HasSuffix(endpoint, ".woff") {
						continue
					}
					mu.Lock()
					discoveredEndpoints[endpoint] = true
					mu.Unlock()
				}
			}
		}(jsURL)
	}
	wg.Wait()

	// Probe discovered endpoints
	if len(crawl.Pages) > 0 && len(discoveredEndpoints) > 0 {
		base := strings.Split(crawl.Pages[0].URL, "/")[0] + "//" + strings.Split(crawl.Pages[0].URL, "/")[2]
		var liveEndpoints []string
		for endpoint := range discoveredEndpoints {
			fullURL := endpoint
			if !strings.HasPrefix(endpoint, "http") {
				fullURL = base + endpoint
			}
			resp := http.Get(fullURL)
			if resp.Err == nil && resp.StatusCode != 404 && resp.StatusCode < 500 {
				liveEndpoints = append(liveEndpoints, fullURL)
			}
		}
		if len(liveEndpoints) > 0 {
			findings = append(findings, Finding{
				Type: "JS Endpoint Discovery", Severity: "info",
				URL:    fmt.Sprintf("%d endpoints from %d JS files", len(liveEndpoints), len(jsURLs)),
				Detail: fmt.Sprintf("Found %d live API endpoints in JavaScript: %s", len(liveEndpoints), strings.Join(liveEndpoints[:min(10, len(liveEndpoints))], ", ")),
				Template: "apex-js-endpoints",
			})
			// Add to crawl data for other scanners to use
			for _, ep := range liveEndpoints {
				parsedURL, _ := url.Parse(ep)
				if parsedURL != nil {
					base := ep
					if parsedURL.RawQuery != "" {
						base = strings.Split(ep, "?")[0]
						for k := range parsedURL.Query() {
							crawl.Params[base] = append(crawl.Params[base], k)
						}
					}
				}
			}
		}
	}
	return findings
}

// --- Auto-Register + Authenticated Scan ---

func scanAuthenticated(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	if len(crawl.Pages) == 0 {
		return findings
	}
	base := strings.Split(crawl.Pages[0].URL, "/")[0] + "//" + strings.Split(crawl.Pages[0].URL, "/")[2]

	// Try to auto-register
	registerPaths := []string{"/register", "/signup", "/api/register", "/api/auth/register", "/api/v1/register", "/api/users"}
	var authToken string

	testEmail := fmt.Sprintf("apex_test_%d@test.com", time.Now().Unix())
	testPass := "ApexTest123!"

	for _, path := range registerPaths {
		body := fmt.Sprintf(`{"email":"%s","password":"%s","username":"apextest%d"}`, testEmail, testPass, time.Now().Unix()%10000)
		resp := http.Post(base+path, "application/json", body)
		if resp.Err == nil && (resp.StatusCode == 200 || resp.StatusCode == 201) {
			// Extract token from response
			if strings.Contains(resp.Body, "token") {
				// Simple token extraction
				for _, part := range strings.Split(resp.Body, "\"") {
					if strings.HasPrefix(part, "eyJ") || (len(part) > 20 && !strings.Contains(part, " ")) {
						authToken = part
						break
					}
				}
			}
			if authToken == "" {
				// Try login
				loginPaths := []string{"/login", "/api/login", "/api/auth/login", "/api/v1/login"}
				for _, lp := range loginPaths {
					loginBody := fmt.Sprintf(`{"email":"%s","password":"%s"}`, testEmail, testPass)
					lr := http.Post(base+lp, "application/json", loginBody)
					if lr.Err == nil && lr.StatusCode == 200 {
						for _, part := range strings.Split(lr.Body, "\"") {
							if strings.HasPrefix(part, "eyJ") || (len(part) > 20 && !strings.Contains(part, " ") && !strings.Contains(part, ":")) {
								authToken = part
								break
							}
						}
						break
					}
				}
			}
			break
		}
	}

	if authToken == "" {
		return findings
	}

	findings = append(findings, Finding{
		Type: "Auto-Registration Successful", Severity: "info",
		URL: base, Detail: fmt.Sprintf("Registered and obtained auth token (len=%d)", len(authToken)),
		Template: "apex-auto-register",
	})

	// Now crawl authenticated endpoints and compare
	authPaths := []string{"/api/me", "/api/user", "/api/profile", "/api/account", "/api/settings", "/api/admin", "/api/users", "/api/dashboard", "/api/orders", "/api/payments", "/api/transactions"}
	for _, path := range authPaths {
		// Unauthenticated
		unauthResp := http.Get(base + path)
		// Authenticated
		items := []engine.RequestItem{{URL: base + path, Method: "GET", Headers: map[string]string{"Authorization": "Bearer " + authToken}}}
		for authResp := range http.BatchRequest(items, 1) {
			if authResp.Err != nil {
				continue
			}
			// If auth gives 200 but unauth gives 401/403 — endpoint exists behind auth
			if authResp.StatusCode == 200 && (unauthResp.StatusCode == 401 || unauthResp.StatusCode == 403) {
				findings = append(findings, Finding{
					Type: "Authenticated Endpoint Found", Severity: "info",
					URL: base + path, Detail: fmt.Sprintf("Returns data with auth (status %d, %d bytes)", authResp.StatusCode, authResp.Size),
					Template: "apex-auth-endpoint",
				})
				// Test IDOR — try accessing other users' data
				if strings.Contains(authResp.Body, "id") {
					// Modify the token or try different IDs
					idPaths := []string{base + path + "/1", base + path + "/2", base + path + "?user_id=1"}
					for _, idPath := range idPaths {
						idItems := []engine.RequestItem{{URL: idPath, Method: "GET", Headers: map[string]string{"Authorization": "Bearer " + authToken}}}
						for idResp := range http.BatchRequest(idItems, 1) {
							if idResp.Err == nil && idResp.StatusCode == 200 && idResp.Body != authResp.Body && len(idResp.Body) > 50 {
								findings = append(findings, Finding{
									Type: "IDOR via Authenticated Endpoint", Severity: "high",
									URL: idPath, Detail: "Different user data accessible with our auth token",
									Template: "apex-auth-idor",
								})
							}
						}
					}
				}
			}
		}
	}
	return findings
}

// --- Arjun-Style Param Brute-Forcer ---

func scanParamBruteforce(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	loadPayloads()
	if len(paramWordlist) == 0 {
		return findings
	}

	var mu sync.Mutex
	sem := make(chan struct{}, cfg.Threads)
	var wg sync.WaitGroup

	// Test on main pages
	targets := []string{}
	for _, page := range crawl.Pages[:min(5, len(crawl.Pages))] {
		targets = append(targets, page.URL)
	}

	for _, target := range targets {
		// Get baseline
		baseResp := http.Get(target)
		if baseResp.Err != nil || baseResp.StatusCode != 200 {
			continue
		}
		baseSize := len(baseResp.Body)
		baseHeaders := fmt.Sprintf("%v", baseResp.Headers)

		// Test params in batches of 10 (faster than one-by-one)
		for i := 0; i < len(paramWordlist); i += 10 {
			batch := paramWordlist[i:min(i+10, len(paramWordlist))]
			wg.Add(1)
			go func(params []string, baseURL string) {
				defer wg.Done()
				sem <- struct{}{}
				defer func() { <-sem }()

				// Build URL with all params in batch
				sep := "?"
				if strings.Contains(baseURL, "?") {
					sep = "&"
				}
				paramStr := ""
				for _, p := range params {
					paramStr += p + "=apex_test&"
				}
				testURL := baseURL + sep + strings.TrimRight(paramStr, "&")
				resp := http.Get(testURL)
				if resp.Err != nil {
					return
				}

				// If response differs, test individually
				if resp.StatusCode == 200 && (len(resp.Body) != baseSize || fmt.Sprintf("%v", resp.Headers) != baseHeaders) {
					for _, p := range params {
						indivURL := baseURL + sep + p + "=apex_test"
						indivResp := http.Get(indivURL)
						if indivResp.Err == nil && indivResp.StatusCode == 200 {
							sizeDiff := len(indivResp.Body) - baseSize
							if sizeDiff > 20 || sizeDiff < -20 || indivResp.Body != baseResp.Body {
								mu.Lock()
								findings = append(findings, Finding{
									Type: "Hidden Parameter: " + p, Severity: "medium",
									URL: indivURL, Param: p,
									Detail:   fmt.Sprintf("Param '%s' changes response (size diff: %+d bytes)", p, sizeDiff),
									Template: "apex-hidden-param",
								})
								// Add to crawl data
								crawl.Params[baseURL] = append(crawl.Params[baseURL], p)
								mu.Unlock()
							}
						}
					}
				}
			}(batch, target)
		}
	}
	wg.Wait()
	return findings
}

// --- Differential Response Analyzer ---

func scanDifferential(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	var mu sync.Mutex
	sem := make(chan struct{}, cfg.Threads)
	var wg sync.WaitGroup

	for _, page := range crawl.Pages[:min(20, len(crawl.Pages))] {
		wg.Add(1)
		go func(pageURL string) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()

			// Baseline: normal request
			baseResp := http.Get(pageURL)
			if baseResp.Err != nil || baseResp.StatusCode != 200 {
				return
			}

			// Test 1: With vs without common auth headers
			authHeaders := []struct{ header, value, name string }{
				{"Authorization", "Bearer test", "auth bearer"},
				{"X-API-Key", "test", "api key"},
				{"Cookie", "admin=true; role=admin", "admin cookie"},
				{"X-Forwarded-For", "127.0.0.1", "internal IP"},
			}
			for _, ah := range authHeaders {
				items := []engine.RequestItem{{URL: pageURL, Method: "GET", Headers: map[string]string{ah.header: ah.value}}}
				for resp := range http.BatchRequest(items, 1) {
					if resp.Err == nil && resp.StatusCode == 200 && resp.Body != baseResp.Body {
						sizeDiff := len(resp.Body) - len(baseResp.Body)
						if sizeDiff > 50 { // More content with header = info leak
							mu.Lock()
							findings = append(findings, Finding{
								Type: "Differential: Extra Data with " + ah.name, Severity: "high",
								URL: pageURL, Detail: fmt.Sprintf("Header '%s: %s' reveals %d extra bytes", ah.header, ah.value, sizeDiff),
								Template: "apex-differential",
							})
							mu.Unlock()
						}
					}
				}
			}

			// Test 2: Different HTTP methods reveal different data
			for _, method := range []string{"POST", "PUT", "PATCH", "DELETE"} {
				items := []engine.RequestItem{{URL: pageURL, Method: method, Headers: map[string]string{"Content-Type": "application/json"}}}
				for resp := range http.BatchRequest(items, 1) {
					if resp.Err == nil && resp.StatusCode == 200 && resp.Body != baseResp.Body && len(resp.Body) > len(baseResp.Body)+50 {
						mu.Lock()
						findings = append(findings, Finding{
							Type: "Differential: " + method + " Returns Extra Data", Severity: "medium",
							URL: pageURL, Detail: fmt.Sprintf("%s returns %d more bytes than GET", method, len(resp.Body)-len(baseResp.Body)),
							Template: "apex-differential-method",
						})
						mu.Unlock()
						break
					}
				}
			}

			// Test 3: Content-Type confusion
			ctTests := []string{"application/json", "application/xml", "text/xml", "application/x-www-form-urlencoded"}
			for _, ct := range ctTests {
				resp := http.Post(pageURL, ct, "{}")
				if resp.Err == nil && resp.StatusCode == 200 && resp.Body != baseResp.Body && len(resp.Body) > 50 {
					if strings.Contains(resp.Body, "error") && (strings.Contains(resp.Body, "stack") || strings.Contains(resp.Body, "trace") || strings.Contains(resp.Body, "debug")) {
						mu.Lock()
						findings = append(findings, Finding{
							Type: "Content-Type Confusion — Debug Info Leak", Severity: "medium",
							URL: pageURL, Detail: fmt.Sprintf("Content-Type '%s' triggers debug/error output", ct),
							Template: "apex-ct-confusion",
						})
						mu.Unlock()
						break
					}
				}
			}
		}(page.URL)
	}
	wg.Wait()
	return findings
}

File: src/scanners/power_upgrades.cpp
/// @file scanners/power_upgrades.cpp
/// @brief Power upgrades: JS endpoint extraction, auto-register + auth scan,
///        param brute-force, differential response analysis.
#include "scanner_base.hpp"
#include "../payloads.hpp"
#include <fstream>
#include <set>
#include <sstream>

namespace apex {
namespace {

/// Load lines from a file, skipping comments and blanks.
std::vector<std::string> load_wordlist(const std::string &filename) {
  std::vector<std::string> lines;
  // Iterate over targets.
  for (const auto &dir : {"./payloads", "../payloads", "/usr/share/apex-cli/payloads"}) {
    std::string path = std::string(dir) + "/" + filename;
    std::ifstream f(path);
    if (!f.is_open()) continue;
    std::string line;
    while (std::getline(f, line)) {
      if (!line.empty() && line[0] != '#') lines.push_back(line);
    }
    break;
  }
  return lines;
}

/// JS endpoint extraction — find API routes in JavaScript files.
/// Scanner implementation.
std::vector<Finding> scan_js_endpoints(const Config &, HttpClient &http,
                                       const CrawlResult &crawl) {
  std::vector<Finding> findings;
  // Early return if no URLs to scan.
  if (crawl.urls.empty()) return findings;
  std::string base = base_url_from(crawl.urls[0]);

  std::set<std::string> js_urls;
  // Iterate over targets.
  for (const auto &url : crawl.urls) {
    if (url.find(".js") != std::string::npos ||
        url.find("/static/js/") != std::string::npos ||
        url.find("/_next/") != std::string::npos)
      js_urls.insert(url);
  }
  // Iterate over targets.
  for (const auto &p : {"/main.js", "/app.js", "/bundle.js", "/vendor.js",
                         "/_next/static/chunks/main.js", "/static/js/main.js"})
    js_urls.insert(base + p);

  std::vector<std::regex> patterns = {
      std::regex(R"(["'](/api/[a-zA-Z0-9_/\-\.]+)["'])"),
      std::regex(R"(["'](/v[0-9]+/[a-zA-Z0-9_/\-\.]+)["'])"),
      std::regex(R"((?:fetch|axios)\s*\(\s*["'`]([^"'`\s]+))"),
      std::regex(R"((?:path|route|endpoint|url)\s*[:=]\s*["']([^"']+)["'])"),
  };

  std::set<std::string> discovered;
  // Iterate over targets.
  for (const auto &js_url : js_urls) {
    auto resp = http.get(js_url);
    if (resp.status_code != 200 || resp.body.size() < 100) continue;
    for (const auto &pat : patterns) {
      auto begin = std::sregex_iterator(resp.body.begin(), resp.body.end(), pat);
      auto end = std::sregex_iterator();
      for (auto it = begin; it != end; ++it) {
        std::string ep = (*it)[1].str();
        if (ep.size() >= 3 && ep.size() <= 200 &&
            ep.find(".css") == std::string::npos &&
            ep.find(".png") == std::string::npos)
          discovered.insert(ep);
      }
    }
  }

  // Probe discovered endpoints.
  std::vector<std::string> live;
  // Iterate over targets.
  for (const auto &ep : discovered) {
    std::string full = ep.find("http") == 0 ? ep : base + ep;
    auto resp = http.get(full);
    if (resp.status_code != 404 && resp.status_code < 500)
      live.push_back(full);
    if (live.size() >= 20) break;
  }

  if (!live.empty()) {
    std::string detail = "Found " + std::to_string(live.size()) +
                         " live API endpoints in JavaScript";
    findings.push_back({"JS Endpoint Discovery", "info",
                        std::to_string(live.size()) + " endpoints from " +
                            std::to_string(js_urls.size()) + " JS files",
                        detail, "", "", ""});
  }
  return findings;
}

/// Auto-register and scan authenticated endpoints.
/// Scanner implementation.
std::vector<Finding> scan_authenticated(const Config &, HttpClient &http,
                                        const CrawlResult &crawl) {
  std::vector<Finding> findings;
  // Early return if no URLs to scan.
  if (crawl.urls.empty()) return findings;
  std::string base = base_url_from(crawl.urls[0]);

  const std::vector<std::string> reg_paths = {
      "/register", "/signup", "/api/register", "/api/auth/register", "/api/v1/register"};

  std::string auth_token;
  std::string test_email = "apex_test_" + std::to_string(time(nullptr)) + "@test.com";

  // Iterate over targets.
  for (const auto &path : reg_paths) {
    std::string body = R"({"email":")" + test_email +
                       R"(","password":"ApexTest123!","username":"apextest"})";
    auto resp = http.post(base + path, body, "application/json");
    if (resp.status_code == 200 || resp.status_code == 201) {
      // Extract token.
      std::regex tok_re(R"(eyJ[A-Za-z0-9_-]{20,})");
      std::smatch m;
      if (std::regex_search(resp.body, m, tok_re)) {
        auth_token = m[0].str();
        break;
      }
    }
  }

  if (auth_token.empty()) return findings;

  findings.push_back({"Auto-Registration Successful", "info", base,
                      "Registered and obtained auth token (len=" +
                          std::to_string(auth_token.size()) + ")",
                      "", "", ""});

  // Probe authenticated endpoints.
  const std::vector<std::string> auth_paths = {
      "/api/me", "/api/user", "/api/profile", "/api/account",
      "/api/admin", "/api/users", "/api/dashboard"};

  // Iterate over targets.
  for (const auto &path : auth_paths) {
    auto unauth = http.get(base + path);
    auto authed = http.get(base + path, {{"Authorization", "Bearer " + auth_token}});
    if (authed.status_code == 200 &&
        (unauth.status_code == 401 || unauth.status_code == 403)) {
      findings.push_back({"Authenticated Endpoint Found", "info", base + path,
                          "Returns data with auth (" + std::to_string(authed.body.size()) + " bytes)",
                          "", "", ""});
      // IDOR check.
      for (const auto &id_path : {path + "/1", path + "/2", path + "?user_id=1"}) {
        auto id_resp = http.get(base + id_path, {{"Authorization", "Bearer " + auth_token}});
        if (id_resp.status_code == 200 && id_resp.body != authed.body &&
            id_resp.body.size() > 50) {
          findings.push_back({"IDOR via Authenticated Endpoint", "high", base + id_path,
                              "Different user data accessible with our auth token",
                              "", "", ""});
        }
      }
    }
  }
  return findings;
}

/// Arjun-style param brute-force.
/// Scanner implementation.
std::vector<Finding> scan_param_bruteforce(const Config &, HttpClient &http,
                                           const CrawlResult &crawl) {
  std::vector<Finding> findings;
  auto params = load_wordlist("params.txt");
  if (params.empty() || crawl.urls.empty()) return findings;

  size_t limit = std::min(crawl.urls.size(), size_t(5));
  for (size_t i = 0; i < limit; ++i) {
    auto base_resp = http.get(crawl.urls[i]);
    if (base_resp.status_code != 200) continue;
    int base_size = (int)base_resp.body.size();

    // Test in batches of 10.
    for (size_t j = 0; j < params.size(); j += 10) {
      std::string param_str;
      size_t end = std::min(j + 10, params.size());
      for (size_t k = j; k < end; ++k)
        param_str += params[k] + "=apex_test&";
      if (!param_str.empty()) param_str.pop_back();

      std::string sep = crawl.urls[i].find('?') != std::string::npos ? "&" : "?";
      auto resp = http.get(crawl.urls[i] + sep + param_str);
      if (resp.status_code != 200) continue;

      int diff = (int)resp.body.size() - base_size;
      if (std::abs(diff) <= 20) continue;

      // Narrow down which param caused it.
      for (size_t k = j; k < end; ++k) {
        auto indiv = http.get(crawl.urls[i] + sep + params[k] + "=apex_test");
        if (indiv.status_code == 200) {
          int idiff = (int)indiv.body.size() - base_size;
          if (std::abs(idiff) > 20) {
            findings.push_back({"Hidden Parameter: " + params[k], "medium",
                                crawl.urls[i] + sep + params[k] + "=apex_test",
                                "Param '" + params[k] + "' changes response (size diff: " +
                                    std::to_string(idiff) + " bytes)",
                                params[k], "", ""});
          }
        }
      }
    }
  }
  return findings;
}

/// Differential response analysis.
/// Scanner implementation.
std::vector<Finding> scan_differential(const Config &, HttpClient &http,
                                       const CrawlResult &crawl) {
  std::vector<Finding> findings;
  size_t limit = std::min(crawl.urls.size(), size_t(20));

  for (size_t i = 0; i < limit; ++i) {
    auto base_resp = http.get(crawl.urls[i]);
    if (base_resp.status_code != 200) continue;

    // Test auth headers that might reveal extra data.
    struct AH { const char *header; const char *value; const char *name; };
    const AH auth_headers[] = {
        {"Authorization", "Bearer test", "auth bearer"},
        {"X-API-Key", "test", "api key"},
        {"Cookie", "admin=true; role=admin", "admin cookie"},
        {"X-Forwarded-For", "127.0.0.1", "internal IP"},
    };

    for (const auto &ah : auth_headers) {
      auto resp = http.get(crawl.urls[i], {{ah.header, ah.value}});
      if (resp.status_code == 200 && resp.body != base_resp.body) {
        int diff = (int)resp.body.size() - (int)base_resp.body.size();
        if (diff > 50) {
          findings.push_back({"Differential: Extra Data with " + std::string(ah.name),
                              "high", crawl.urls[i],
                              std::string("Header '") + ah.header + ": " + ah.value +
                                  "' reveals " + std::to_string(diff) + " extra bytes",
                              "", "", ""});
        }
      }
    }

    // Test different HTTP methods.
    for (const auto &method : {"POST", "PUT", "PATCH", "DELETE"}) {
      auto resp = http.post(crawl.urls[i], "{}", "application/json");
      if (resp.status_code == 200 && resp.body != base_resp.body &&
          (int)resp.body.size() > (int)base_resp.body.size() + 50) {
        findings.push_back({"Differential: " + std::string(method) + " Returns Extra Data",
                            "medium", crawl.urls[i],
                            std::string(method) + " returns " +
                                std::to_string(resp.body.size() - base_resp.body.size()) +
                                " more bytes than GET",
                            "", "", ""});
        break;
      }
    }

    // Content-Type confusion.
    for (const auto &ct : {"application/json", "application/xml", "text/xml"}) {
      auto resp = http.post(crawl.urls[i], "{}", ct);
      if (resp.status_code == 200 && resp.body != base_resp.body &&
          (resp.body.find("stack") != std::string::npos ||
           resp.body.find("trace") != std::string::npos ||
           resp.body.find("debug") != std::string::npos)) {
        findings.push_back({"Content-Type Confusion — Debug Info Leak", "medium",
                            crawl.urls[i],
                            std::string("Content-Type '") + ct + "' triggers debug output",
                            "", "", ""});
        break;
      }
    }
  }
  return findings;
}

} // namespace

std::vector<Scanner> register_power_scanners() {
  return {
      {"JS Endpoints", scan_js_endpoints},
      {"Authenticated Scan", scan_authenticated},
      {"Param Brute-Force", scan_param_bruteforce},
      {"Differential Analysis", scan_differential},
  };
}

} // namespace apex
